app/src/main/kotlin/trees/BSTree.kt

```diff
@@ -0,0 +1,47 @@
+package trees
+
+import trees.abstract_trees.BinaryTree
+import trees.nodes.BSNode
+
+class BSTree<K : Comparable<K>, V> : BinaryTree<K, V, BSNode<K, V>>() {
+
+    override fun add(key: K, value: V) {
+        root = recursiveAdd(root, BSNode(key, value))
+    }
+
+    private fun recursiveAdd(currentNode: BSNode<K, V>?, node: BSNode<K, V>): BSNode<K, V> {
+        if (currentNode == null) return node
+        if (currentNode.key < node.key) {
+            currentNode.right = recursiveAdd(currentNode.right, node)
+        }
+        if (currentNode.key > node.key) {
+            currentNode.left = recursiveAdd(currentNode.left, node)
+        }
+        if (currentNode.key == node.key) {
+            return node
+        }
+        return currentNode
+    }
+
+    override fun remove(key: K) {
+        val node = find(key) ?: throw IllegalStateException("There is no such node with key $key in the tree")
```

Comment:
Во-первых, эта ошибка из-за того, что вам передали невалидный аргумент, то есть надо было использовать `IllegalArgumentException`.
Во-вторых, вроде бы вам уже это писали, но обычно проверки с `IllegalStateException` и `IllegalArgumentException` компактнее делать через `check` и `require` соответственно (а проверки на null -- через `checkNotNull` и `requireNotNull`:
```suggestion
        val node = requireNotNull(find(key)) { "There is no such node with key $key in the tree" }
```